protocol/nats_jetstream/v3/message_test.go

```diff
@@ -190,3 +198,116 @@ func TestGetExtension(t *testing.T) {
 		})
 	}
 }
+
+func TestFinish(t *testing.T) {
+	type args struct {
+		err    error
+		ackErr error
+		nakErr error
+	}
+	type wants struct {
+		err        error
+		ackCalled  bool
+		nackCalled bool
+	}
+	tests := []struct {
+		name  string
+		args  args
+		wants wants
+	}{
+		{
+			name: "nil error given",
+			args: args{
+				err: nil,
+			},
+			wants: wants{
+				err:        nil,
+				ackCalled:  true,
+				nackCalled: false,
+			},
+		},
+		{
+			name: "ACK error given",
+			args: args{
+				err: protocol.ResultACK,
+			},
+			wants: wants{
+				err:        nil,
+				ackCalled:  true,
+				nackCalled: false,
+			},
+		},
+		{
+			name: "NACK error given",
+			args: args{
+				err: protocol.ResultNACK,
+			},
+			wants: wants{
+				err:        nil,
+				ackCalled:  false,
+				nackCalled: true,
+			},
+		},
+		{
+			name: "unknown error given",
+			args: args{
+				err: errors.New("unknown"),
+			},
+			wants: wants{
+				err:        nil,
+				ackCalled:  false,
+				nackCalled: false,
+			},
+		},
+		{
+			name: "jetstream.ErrMsgAlreadyAckd error returned from Ack",
+			args: args{
+				err:    protocol.ResultACK,
+				ackErr: jetstream.ErrMsgAlreadyAckd,
+			},
+			wants: wants{
+				err:        nil,
+				ackCalled:  true,
+				nackCalled: false,
+			},
+		},
+		{
+			name: "jetstream.ErrMsgAlreadyAckd error returned from Nak",
```

Comment:
Interesting, there's a case when a (protocol) NACK would contain NATS ACK-ed in this implementation?

Reply:
nak and ack make the same NATS call deep down with a minor difference in the ackType argument passed via the API.  The ErrMsgAlreadyAckd error that comes back simply means that the message has been already "marked" as having been processed.  It is used for both ack and nak.

In practice, a message should not be marked ack or nak before calling Finish.  But there may be cases where a binding may have a behavior to handle MalformedEvent or a responder may be interjected.  In these cases, I believe Finish will still only called once.  The error check is simply there to not send back an error in the unlikely case somehow Finish is called more than once.